main.cpp

```diff
@@ -28,57 +28,57 @@ struct T
 
 struct Struct1                                //4
 {
-    T* compare(T* a, T* b) //5
+    const T* compare(const T& a,const T& b) //5
     {
-        if(a != nullptr && b != nullptr) 
-        {
-            if( a->value < b->value ) return a;
-            if( a->value > b->value ) return b;
-        }
-        return nullptr;
+        if( a.value < b.value ) return &a;
+        if( a.value > b.value ) return &a;
+        return &a;   // Added to remove the compiler warnings "control may reach end of non-void function [-Wreturn-type]"
+    }
+
+    T* compare2(T& a, T& b) //5
+    {
+        if( a.value < b.value ) return &a;
+        if( a.value > b.value ) return &a;
+        return &a;   // Added to remove the compiler warnings "control may reach end of non-void function [-Wreturn-type]"
     }
```

Comment:
You have changed the internal logic of the function and no longer return `a`,`b`, or `nullptr`.
You need to revert the code back to the original version and then try to convert whatever you can into references.

When you try to replace the pointers in this function with references, ask yourself:
- can **the return type** `T*` be converted to a reference? hint: you're returning a `nullptr` at the end.
- can `T* a` be converted to a reference?
- can `T* b` be converted to a reference?

Revise the code based on your answers to those questions.

- [ ] revise

Reply:
> can the return type T* be converted to a reference? hint: you're returning a nullptr at the end.
yes, it's possible because in the scope of the member function exists valid references of a and b
> can T* a be converted to a reference?
yes, it's possible because the member function can get the object t1 via function parameter 
> can T* b be converted to a reference?
yes, it's possible because the member function can get the object t2 via function parameter